quipucords/api/credential/test_serializer.py

```diff
@@ -151,3 +218,298 @@ def test_change_with_same_cred_type():
+        """Test validation of a payload without password and ssh_key."""
+        serializer = CredentialSerializer(
+            data={
+                "name": "network-credential",
+                "cred_type": DataSources.NETWORK,
+                "username": "john-doe",
+            }
+        )
+        assert not serializer.is_valid()
+        assert serializer.errors == {"non_field_errors": [messages.HC_PWD_OR_KEYFILE]}
+
+    @pytest.mark.django_db
+    @pytest.mark.xfail(
+        reason="This behavior is not implemented yet - clients are forced to do it."
+    )
+    def test_replace_password(self, credential_with_ssh_key):
+        """Test replacing password with ssh_keyfile."""
+        serializer = CredentialSerializer(
+            data={
+                "password": "super-duper-secret",
+            },
+            instance=credential_with_ssh_key,
+        )
+        assert serializer.is_valid(), serializer.errors
+        serializer.save()
+        assert (
+            decrypt_data_as_unicode(credential_with_ssh_key.password)
+            == "super-duper-secret"
+        )
+        assert credential_with_ssh_key.ssh_keyfile is None
+
+    @pytest.mark.django_db
+    @pytest.mark.xfail(
+        reason="This behavior is not implemented yet - clients are forced to do it."
+    )
+    def test_replace_ssh_keyfile(self, credential_with_password, fake_ssh_key):
+        """Test replacing ssh_keyfile with password."""
+        serializer = CredentialSerializer(
+            data={
+                "ssh_keyfile": fake_ssh_key,
+            },
+            instance=credential_with_password,
+        )
+        assert serializer.is_valid(), serializer.errors
+        serializer.save()
+        assert credential_with_password.password is None
+        assert credential_with_password.ssh_keyfile == fake_ssh_key
+
+    @pytest.fixture(scope="class")
+    def payload_with_passphrase_without_ssh_key(self):
+        """Return a payload with passphrase and without ssh key."""
+        return {
+            "name": "network-credential",
+            "cred_type": DataSources.NETWORK,
+            "username": "john-doe",
+            "password": "shhhhhhhhhh-this-is-secret",
+            "ssh_passphrase": "another-secret",
+        }
+
+    def test_create_ssh_passphrase_without_ssh_keyfile(
+        self, payload_with_passphrase_without_ssh_key
+    ):
+        """Test using ssh_passphrase without a ssh key."""
+        serializer = CredentialSerializer(
+            data=payload_with_passphrase_without_ssh_key,
+        )
+        assert not serializer.is_valid()
+        assert serializer.errors == {"ssh_passphrase": [messages.HC_NO_KEY_W_PASS]}
+
+    @pytest.mark.django_db
+    @pytest.mark.parametrize(
+        "extra_serializer_kwargs",
+        (
+            pytest.param({}, id="full-update"),
+            pytest.param(
+                {"partial": True},
+                marks=pytest.mark.xfail(reason="BUG DETECTED!"),
```

Comment:
this (and the other xfailed tests) will be fixed in a follow up PR

Reply:
just by the sheer size the PR. also because these bugs were not introduced here.

If you don't mind I can push more commits fixing it.

Reply:
> also because these bugs were not introduced here.
Ah! I did not realize these were preexisting issues. In that case, sure, I leave it at your discretion!